modules/meshrenderinggl/include/modules/meshrenderinggl/datastructures/rasterization.h

```diff
@@ -0,0 +1,79 @@
+/*********************************************************************************
+ *
+ * Inviwo - Interactive Visualization Workshop
+ *
+ * Copyright (c) 2019-2020 Inviwo Foundation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#pragma once
+
+#include <modules/meshrenderinggl/meshrenderingglmoduledefine.h>
+#include <modules/opengl/shader/shaderutils.h>
+
+#include <vector>
+#include <unordered_map>
+#include <optional>
+
+namespace inviwo {
+
+/**
+ * \brief A functor class for rendering geometry into a fragment list
+ * Will be applied by a renderer containing an A-buffer.
+ */
+
+class IVW_MODULE_MESHRENDERINGGL_API Rasterization {
+public:
+    virtual ~Rasterization() = default;
+
+    /**
+     * \brief Render the fragments, with all setup and evaluation taken care of.
+     * If opaque is set, a standard render call instead.
+     * @param imageSize Size in pixels.
+     * @param setUniforms Binds the fragment list buffer and sets required uniforms.
+     */
+    virtual void rasterize(const ivec2& imageSize,
+                           std::function<void(Shader&)> setUniforms) const = 0;
+
+    /**
+     * \brief Query whether fragments will be emitted.
+     * @return True for order-independent rendering, false for opaque rasterization.
+     */
+    virtual bool usesFragmentLists() const = 0;
+
+    /**
+     * \brief Get data description for the network interface.
+     * @return Specific information about this rasterization type/instance.
+     */
+    virtual Document getDescription() const;

```

Comment:
Is is possible to move this outside of the class? Printing info is not really its reponsibility

Reply:
But it is, se above. how else would you describe what's being rendered?

Reply:
but we should name it getInfo to be consistent with the other data structures https://github.com/inviwo/inviwo/blob/6068897cfeb1a10a7c5720595f7c6d50b5c6cd02/include/inviwo/core/datastructures/volume/volume.h#L79

Reply:
the info is used here
https://github.com/inviwo/inviwo/blob/6068897cfeb1a10a7c5720595f7c6d50b5c6cd02/include/inviwo/core/ports/datainport.h#L267

the default DataTraits implementation will use
https://github.com/inviwo/inviwo/blob/6068897cfeb1a10a7c5720595f7c6d50b5c6cd02/include/inviwo/core/datastructures/datatraits.h#L103

to get the `getInfo` with help from
https://github.com/inviwo/inviwo/blob/6068897cfeb1a10a7c5720595f7c6d50b5c6cd02/include/inviwo/core/util/introspection.h#L199

So we could add a specialisation to the DataTraits class to do this, but then that specialisation would still need to be able to get the information out of the rasteriser some how...

Reply:
Yes, I think that it is the responsibility of the DataTrait to provide information. The ones in Volume and other data structures are there due to legacy (we did not have DataTraits at the time)...
Having the getInfo in the class breaks the single responsibility principle

Reply:
but then you would have to have a way of exposing what's "inside" rasersize any way. And since what is in it depends on which derived class it is, it has to be a virtual function in the base class. Now what should that function return? I can be lots of different thing in a rasterizer, meshes, textures, shaders, etc? hence it probably has to return some kind of generic string description. hence we get what we have now... The reason for having DataTraits it so we can allow for decorating data with info also when we can't add the getInfo method into the data class, like for a std::vector for example.

Reply:
One trait per derived class would do it. Anyway, I guess convenience wins here

Reply:
You can only specialise for the base class not the derived ones, since the port only know knows about the base. Hence that will not work.

Reply:
True, marked it as resolved